drivers/net/ethernet/intel/idpf/idpf_eth_idc.c

```diff
@@ -0,0 +1,453 @@
+	eth_dev->eth_info.caps.rss_caps = adapter->caps.rss_caps;
+	eth_dev->eth_info.caps.other_caps = adapter->caps.other_caps;
+	eth_dev->eth_info.caps.min_sso_packet_len =
+		adapter->caps.min_sso_packet_len;
+	eth_dev->eth_info.caps.max_sg_bufs_per_tx_pkt =
+		adapter->caps.max_sg_bufs_per_tx_pkt;
+}
+
+/**
+ * idpf_eth_idc_device_alloc - allocate auxiliary device
+ * @adapter: Idpf private structure
+ * @vport_type: Vport type
+ *
+ * Returns allocated structure
+ */
+static struct idpf_eth_idc_auxiliary_dev *
+idpf_eth_idc_device_alloc(struct idpf_adapter *adapter,
+			  enum idpf_vport_type vport_type)
+{
+	struct idpf_eth_idc_auxiliary_dev *new_eth_dev;
+	struct idpf_eth_idc_dev_info *dev_info;
+	enum idpf_queue_type queue_type;
+	u16 i, err;
+
+	for (i = 0; i < adapter->default_vports; ++i) {
+		/* Check pre-allocated one first */
+		if (adapter->adevs[i])
+			continue;
+
+		adapter->adevs[i] =
+			kzalloc(sizeof(struct idpf_eth_idc_auxiliary_dev),
+				GFP_KERNEL);
+		new_eth_dev = adapter->adevs[i];
+		if (!new_eth_dev)
+			goto alloc_exit;
+
+		new_eth_dev->adev.id = ida_alloc(&idpf_eth_idc_ida, GFP_KERNEL);
+		new_eth_dev->eth_info.eth_shared = &adapter->eth_shared;
+		new_eth_dev->eth_info.idx = i;
+
+		/* Set vport type */
+		new_eth_dev->eth_info.vport_type = vport_type;
+
+		if (vport_type == IDPF_DEFAULT_VPORT)
+			queue_type = IDPF_QUEUE_DEFAULT_VPORT;
+		else
+			queue_type = IDPF_QUEUE_DYNAMIC_VPORT;
+
+		dev_info = &new_eth_dev->eth_info;
+		err = idpf_alloc_max_qs(adapter, &dev_info->caps.q_info,
+					queue_type);
+		if (err) {
+			kfree(adapter->adevs[i]);
+			adapter->adevs[i] = NULL;
+			new_eth_dev = NULL;
+		}
+
+		return new_eth_dev;
+	}
+
+alloc_exit:
+	return NULL;
+}
+
+/**
+ * idpf_eth_idc_device_init - Initialize ethernet IDC & device(s)
+ * @adapter: Idpf private structure
+ */
+void idpf_eth_idc_device_init(struct idpf_adapter *adapter)
+{
+	struct device *dev = &adapter->eth_shared.pdev->dev;
+	struct idpf_eth_idc_auxiliary_driver *eth_idc_drv;
+	struct idpf_eth_idc_auxiliary_dev *eth_dev;
+	int err;
+	u16 i;
+
+	if (adapter->adevs) {
+		/* This is reset case so simply invoke add explicitly.
+		 * Direct eth device re-add or probe
+		 */
+		eth_idc_drv = idpf_eth_idc_get_driver();
+		for (i = 0; i < adapter->default_vports; ++i) {
+			eth_dev = adapter->adevs[i];
+			if (!eth_dev)
+				continue;
+
+			eth_idc_drv->adrv.probe(&eth_dev->adev, NULL);
+		}
+		idpf_eth_idc_dispatch_event(adapter,
+					    IDPF_ETH_IDC_EVENT_ALL_VPORTS,
+					    IDPF_ETH_IDC_EVENT_PROBE_POST_INIT,
+					    NULL);
+		return;
+	}
+
+	/* Following is the case of initial probe */
+	adapter->adevs =
+		kzalloc((adapter->max_vports *
+			 sizeof(struct idpf_eth_idc_auxiliary_dev *)),
+			GFP_KERNEL);
+	if (!adapter->adevs)
+		return;
+
+	for (i = 0; i < adapter->default_vports; ++i) {
+		eth_dev = idpf_eth_idc_device_alloc(adapter,
+						    IDPF_DEFAULT_VPORT);
+		if (!eth_dev)
+			return;
+		eth_dev->adev.name = "eth";
```

Comment:
These three variables can also be moved to idc_device_alloc function to have all the initialization at one place. Will look cleaner.

Reply:
Done!